File: LeetCode/golang/0128.go
func longestConsecutive(nums []int) int {
	/**
	* Time:  O(n)
	* Space: O(n)
	*/
	mem := make(map[int]struct{})  // empty struct is a value (no overhead)
    for _, num := range nums {
        mem[num] = struct{}{}
    }

    maxLength := 0
    for num, _ := range mem {
        if _, exists := mem[num - 1]; !exists { // check if not in the beginning of sequence
            length := 0
            
            for { // while has neighbors to right
                if _, exists := mem[num + length]; !exists {
                    break
                }
                length++
            }

            maxLength = max(maxLength, length)
        }
    }

    return maxLength
}

File: LeetCode/python/0128.py
class Node:
    def __init__(self, val=None):
        self.prev = None
        self.next = None
        self.val = val


class Solution:
    def longestConsecutive(self, nums: list[int]) -> int:
        """
        Time:  O(n)
        Space: O(n)
        """
        mem = {}
        for num in nums:
            if num not in mem:
                mem[num] = Node(num)

                left_num, right_num = num - 1, num + 1
                if left_num in mem:
                    mem[left_num].next = mem[num]
                    mem[num].prev = mem[left_num]
                if right_num in mem:
                    mem[right_num].prev = mem[num]
                    mem[num].next = mem[right_num]

        max_counter = 0
        for node in list(mem.values()):
            if node.prev is None:
                counter = 0
                ptr = node
                while ptr:
                    counter += 1
                    ptr = ptr.next
                max_counter = max(max_counter, counter)

        return max_counter
